docs/strings.md

```diff
@@ -0,0 +1,464 @@
+ */
+typedef Unichar = U32;
+
+// Placeholder until we have syntax for value types.
+value Range
+{
+	U64 index;
+	U64 length;
+}
+
+/**
+ * Interface for string types.  Defines only the primitive operations that all
+ * strings must implement.
+ */
+interface String
+{
+	/**
+	 * Return the length of the string in Unicode code points.
+	 */
+	func length(self: readonly & String) : size_t;
+
+	/**
+	 * The encoding that this string can most easily convert to.  This will
+	 * typically be the encoding that the string uses internally.
+	 * Note that a return value of `StringEncodingASCII` implies
+	 * `StringEncodingUTF8` for all read operations because ASCII is a strict
+	 * superset of UTF-8.
+	 */
+	func fastest_encoding(self: readonly & String) : StringEncoding;
+
+	/**
+	 * The standard encoding that this string can most easily convert to.  This
+	 * will typically be the encoding that the string uses internally but may
+	 * not be for custom encodings.  For all standard-library string types,
+	 * this will return the same value as `fastest_encoding`.
+	 */
+	func fastest_unicode_encoding(self: readonly & String) : UnicodeStringEncoding;
+
+	/**
+	 * The standard encoding that will use the smallest amount of space to
+	 * store the characters.  For example, most English strings will return
+	 * `StringEncodingASCII`, French will typically return
+	 * `StringEncodingUTF8`, whereas CJK language strings will likely return
+	 * `StringEncodingUTF16`.
+	 */
+	func smallest_unicode_encoding(self: readonly & String) : UnicodeStringEncoding;
+
+	/**
+	 * Access a specific character.  Returns a unicode code point for the
+	 * specific index in the string.
+	 */
+	func apply(self: readonly & String, index: U64) : Unichar;
+
+	/**
+	 * Access a specific character.
+	 */
+	func apply(self: readonly & String, index: U64) : Unichar;
+
+	/**
+	 * Takes a range in the string (expressed in Unicode code points) and
+	 * optionally a buffer.  Either the desired characters (in the specified
+	 * encoding) are copied into that buffer, or returns an immutable array
+	 * that may point to the internal storage for the string.
+	 *
+	 * If the caller first calls `fastest_unicode_encoding` and uses the result
+	 * as the generic parameter then it will return an internal buffer for
+	 * immutable strings.  This is intended to be used to build fast iterators.
+	 */
+	func copy_or_view_data[StringEncoding Enc](self: readonly & String,
+	                                           range: Range,
+	                                           buffer: Optional[Array[Enc.CharType] & mut])
+	     : Array[Enc.CharType] & (mut | imm);
+
+	/******************************************************************************
+	 * Operations on mutable strings
+	 *****************************************************************************/
+
+	/**
+	 * Updates a specific character. 
+	 */
+	func update(self: (mut | iso) & String, index: U64, c: Unichar);
```

Comment:
I think maybe:

```
update(self: mut, index: USize, c: Unichar): Unichar
```

If `self` is `iso` at the call-site, we'll be able to use it here as `mut` anyway. If we take something that may be `iso`, we can't let the call-site keep it! Returning the previous value at that index, just because that's convenient for most `update` calls.